File: dbus_msg.py
"""Minimal D-Bus message implementation — replaces jeepney for generate_payload.py."""

import struct

MESSAGE_TYPE_METHOD_CALL = 1

FIELD_PATH = 1
FIELD_INTERFACE = 2
FIELD_MEMBER = 3
FIELD_DESTINATION = 6
FIELD_SIGNATURE = 8


class DBusAddress:
    def __init__(self, *, object_path, bus_name, interface):
        self.object_path = object_path
        self.bus_name = bus_name
        self.interface = interface


class Header:
    def __init__(self, fields):
        self.fields = fields


class Message:
    def __init__(self, header, body, message_type):
        self.header = header
        self.body = body
        self.message_type = message_type

    def serialise(self, serial=1):
        return _serialise(self, serial)

    @classmethod
    def from_buffer(cls, data):
        return _parse(data)


def new_method_call(address, method, signature=None, body=None):
    fields = {}
    fields[FIELD_PATH] = address.object_path
    fields[FIELD_DESTINATION] = address.bus_name
    if address.interface is not None:
        fields[FIELD_INTERFACE] = address.interface
    fields[FIELD_MEMBER] = method
    if signature:
        fields[FIELD_SIGNATURE] = signature
    if body is None:
        body = ()
    elif not isinstance(body, tuple):
        body = (body,)
    return Message(Header(fields), body, MESSAGE_TYPE_METHOD_CALL)


_TYPE_ALIGN = {
    'y': 1, 'b': 4, 'n': 2, 'q': 2, 'i': 4, 'u': 4,
    'x': 8, 't': 8, 'd': 8, 's': 4, 'o': 4, 'g': 1,
    'v': 1, 'h': 4,
}


def _align(offset, a):
    return (offset + a - 1) & ~(a - 1)


def _align_buf(buf, a):
    pos = len(buf)
    aligned = _align(pos, a)
    if aligned > pos:
        buf.extend(b'\x00' * (aligned - pos))


def _pack_string(buf, s):
    data = s.encode('utf-8')
    buf.extend(struct.pack('<I', len(data)))
    buf.extend(data)
    buf.append(0)


def _pack_sig(buf, sig):
    encoded = sig.encode('ascii')
    buf.append(len(encoded))
    buf.extend(encoded)
    buf.append(0)


def _sig_align(sig):
    if not sig:
        return 1
    c = sig[0]
    if c == 'a':
        if len(sig) > 1 and sig[1] == '{':
            return 8
        elem_align = _sig_align(sig[1:])
        return max(4, elem_align)
    elif c == '(':
        inner = sig[1:-1]
        if not inner:
            return 1
        types = _parse_sigs(inner)
        return max((_sig_align(t) for t in types), default=1)
    elif c in _TYPE_ALIGN:
        return _TYPE_ALIGN[c]
    return 1


def _parse_sigs(sig):
    types = []
    i = 0
    while i < len(sig):
        c = sig[i]
        if c == 'a':
            if i + 1 < len(sig) and sig[i + 1] == '{':
                end = sig.index('}', i + 2)
                types.append(sig[i:end + 1])
                i = end + 1
            else:
                types.append(c + sig[i + 1])
                i += 2
        elif c == '(':
            end = sig.index(')', i + 1)
            types.append(sig[i:end + 1])
            i = end + 1
        else:
            types.append(c)
            i += 1
    return types


def _pack_value(buf, sig, value):
    c = sig[0] if sig else ''
    if c == 'y':
        buf.append(value)
    elif c == 'b':
        buf.extend(struct.pack('<I', 1 if value else 0))
    elif c == 'n':
        buf.extend(struct.pack('<h', value))
    elif c == 'q':
        buf.extend(struct.pack('<H', value))
    elif c == 'i':
        buf.extend(struct.pack('<i', value))
    elif c == 'u':
        buf.extend(struct.pack('<I', value))
    elif c == 'x':
        buf.extend(struct.pack('<q', value))
    elif c == 't':
        buf.extend(struct.pack('<Q', value))
    elif c == 'd':
        buf.extend(struct.pack('<d', value))
    elif c in ('s', 'o'):
        _pack_string(buf, value)
    elif c == 'g':
        _pack_sig(buf, value)
    elif c == 'h':
        buf.extend(struct.pack('<I', value))
    elif c == 'v':
        inner_sig, inner_val = value
        _pack_sig(buf, inner_sig)
        _align_buf(buf, _sig_align(inner_sig))
        _pack_value(buf, inner_sig, inner_val)
    elif c == 'a':
        if len(sig) > 1 and sig[1] == '{':
            key_sig = sig[2]
            value_sig = sig[3:-1]
            _align_buf(buf, 4)
            items = list(value.items()) if isinstance(value, dict) else list(value)
            buf.extend(struct.pack('<I', len(items)))
            _align_buf(buf, 8)
            for k, v in items:
                _align_buf(buf, 8)
                _pack_value(buf, key_sig, k)
                _pack_value(buf, 'v', (value_sig, v))
        else:
            elem_sig = sig[1:]
            _align_buf(buf, 4)
            buf.extend(struct.pack('<I', len(value)))
            _align_buf(buf, _sig_align(elem_sig))
            for item in value:
                _pack_value(buf, elem_sig, item)
    elif c == '(':
        inner_sigs = _parse_sigs(sig[1:-1])
        for inner_sig, item in zip(inner_sigs, value):
            _align_buf(buf, _sig_align(inner_sig))
            _pack_value(buf, inner_sig, item)
    else:
        raise ValueError(f"Unknown type: {sig}")


_FIELD_ORDER = [FIELD_PATH, FIELD_INTERFACE, FIELD_MEMBER, FIELD_DESTINATION, FIELD_SIGNATURE]
_FIELD_TYPE = {
    FIELD_PATH: 'o',
    FIELD_INTERFACE: 's',
    FIELD_MEMBER: 's',
    FIELD_DESTINATION: 's',
    FIELD_SIGNATURE: 'g',
}


def _serialise(msg, serial):
    body_buf = bytearray()
    body_sig = msg.header.fields.get(FIELD_SIGNATURE, '')
    if body_sig:
        _pack_value(body_buf, '(' + body_sig + ')', msg.body)

    hdr_buf = bytearray()
    first = True
    for code in _FIELD_ORDER:
        if code in msg.header.fields:
            if not first:
                _align_buf(hdr_buf, 8)
            first = False
            _pack_header_field(hdr_buf, code, _FIELD_TYPE[code], msg.header.fields[code])

    buf = bytearray()
    buf.append(ord('l'))
    buf.append(msg.message_type)
    buf.append(0)
    buf.append(1)
    buf.extend(struct.pack('<I', len(body_buf)))
    buf.extend(struct.pack('<I', serial))
    buf.extend(struct.pack('<I', len(hdr_buf)))
    buf.extend(hdr_buf)
    _align_buf(buf, 8)
    buf.extend(body_buf)
    return bytes(buf)


def _pack_header_field(buf, code, sig, value):
    buf.append(code)
    _pack_sig(buf, sig)
    _align_buf(buf, _sig_align(sig))
    _pack_value(buf, sig, value)


def _parse_single_value(data, offset, sig):
    c = sig[0] if sig else ''
    a = _sig_align(sig)
    offset = _align(offset, a)

    if c == 'y':
        return data[offset], offset + 1
    elif c == 'b':
        return bool(struct.unpack_from('<I', data, offset)[0]), offset + 4
    elif c == 'n':
        return struct.unpack_from('<h', data, offset)[0], offset + 2
    elif c == 'q':
        return struct.unpack_from('<H', data, offset)[0], offset + 2
    elif c == 'i':
        return struct.unpack_from('<i', data, offset)[0], offset + 4
    elif c == 'u':
        return struct.unpack_from('<I', data, offset)[0], offset + 4
    elif c == 'x':
        return struct.unpack_from('<q', data, offset)[0], offset + 8
    elif c == 't':
        return struct.unpack_from('<Q', data, offset)[0], offset + 8
    elif c == 'd':
        return struct.unpack_from('<d', data, offset)[0], offset + 8
    elif c in ('s', 'o'):
        length = struct.unpack_from('<I', data, offset)[0]
        start = offset + 4
        val = data[start:start + length].decode('utf-8')
        return val, start + length + 1
    elif c == 'g':
        length = data[offset]
        start = offset + 1
        val = data[start:start + length].decode('ascii')
        return val, start + length + 1
    elif c == 'h':
        return struct.unpack_from('<I', data, offset)[0], offset + 4
    elif c == 'v':
        sig_len = data[offset]
        inner_sig = data[offset + 1:offset + 1 + sig_len].decode('ascii')
        offset2 = offset + 1 + sig_len + 1
        value, offset2 = _parse_single_value(data, offset2, inner_sig)
        return (inner_sig, value), offset2
    elif c == 'a':
        length = struct.unpack_from('<I', data, offset)[0]
        offset += 4
        if len(sig) > 1 and sig[1] == '{':
            key_sig = sig[2]
            value_sig = sig[3:-1]
            result = []
            offset = _align(offset, 8)
            for _ in range(length):
                offset = _align(offset, 8)
                k, offset = _parse_single_value(data, offset, key_sig)
                v, offset = _parse_single_value(data, offset, 'v')
                result.append((k, v))
            return dict(result), offset
        else:
            elem_sig = sig[1:]
            elem_align = _sig_align(elem_sig)
            offset = _align(offset, elem_align)
            result = []
            for _ in range(length):
                v, offset = _parse_single_value(data, offset, elem_sig)
                result.append(v)
            return result, offset
    elif c == '(':
        inner_sigs = _parse_sigs(sig[1:-1])
        result = []
        for inner_sig in inner_sigs:
            offset = _align(offset, _sig_align(inner_sig))
            v, offset = _parse_single_value(data, offset, inner_sig)
            result.append(v)
        return tuple(result), offset
    else:
        raise ValueError(f"Unknown type: {sig}")


def _parse(data):
    endian = chr(data[0])
    if endian != 'l':
        raise ValueError(f"Unsupported endianness: {endian}")
    message_type = data[1]
    body_len = struct.unpack_from('<I', data, 4)[0]
    hdr_len = struct.unpack_from('<I', data, 12)[0]
    hdr_start = 16
    hdr_end = hdr_start + hdr_len
    hdr_data = data[hdr_start:hdr_end]

    fields = {}
    offset = 0
    while offset < len(hdr_data):
        code = hdr_data[offset]
        offset += 1
        value, offset = _parse_single_value(hdr_data, offset, 'v')
        fields[code] = value[1]
        offset = _align(offset, 8)

    body_start = _align(hdr_end, 8)
    body_end = body_start + body_len
    body_data = data[body_start:body_end]

    body_sig = fields.get(FIELD_SIGNATURE, '')
    body = ()
    if body_sig:
        sig = '(' + body_sig + ')'
        body, _ = _parse_single_value(data, body_start, sig)

    return Message(Header(fields), body, message_type)

File: dbus_msg.zig
const std = @import("std");

pub const MESSAGE_TYPE_METHOD_CALL = 1;

pub const FIELD_PATH = 1;
pub const FIELD_INTERFACE = 2;
pub const FIELD_MEMBER = 3;
pub const FIELD_DESTINATION = 6;
pub const FIELD_SIGNATURE = 8;

const MAX_SIZE = 512;

pub const Buf = struct {
    data: [MAX_SIZE]u8 = undefined,
    off: usize = 0,

    pub fn get(self: *const Buf) []const u8 {
        return self.data[0..self.off];
    }

    pub fn putU8(self: *Buf, v: u8) void {
        self.data[self.off] = v;
        self.off += 1;
    }

    pub fn padTo(self: *Buf, a: usize) void {
        const new_off = alignUp(self.off, a);
        while (self.off < new_off) : (self.off += 1) {
            self.data[self.off] = 0;
        }
    }

    pub fn putU32(self: *Buf, v: u32) void {
        self.padTo(4);
        std.mem.writeInt(u32, self.data[self.off..][0..4], v, .little);
        self.off += 4;
    }

    pub fn putI32(self: *Buf, v: i32) void {
        self.padTo(4);
        std.mem.writeInt(i32, self.data[self.off..][0..4], v, .little);
        self.off += 4;
    }

    pub fn putString(self: *Buf, s: []const u8) void {
        self.padTo(4);
        self.putU32(@intCast(s.len));
        @memcpy(self.data[self.off..][0..s.len], s);
        self.off += s.len;
        self.data[self.off] = 0;
        self.off += 1;
    }

    pub fn putSignature(self: *Buf, s: []const u8) void {
        self.data[self.off] = @intCast(s.len);
        self.off += 1;
        @memcpy(self.data[self.off..][0..s.len], s);
        self.off += s.len;
        self.data[self.off] = 0;
        self.off += 1;
    }

    fn putVariantValue(self: *Buf, comptime sig: []const u8, val: anytype) void {
        self.putSignature(sig);
        self.padTo(sigAlign(sig));
        self.putValue(sig, val);
    }

    fn putArray(self: *Buf, comptime elem_sig: []const u8, items: anytype) void {
        self.padTo(4);
        self.putU32(@intCast(items.len));
        self.padTo(sigAlign(elem_sig));
        inline for (items) |item| {
            self.putValue(elem_sig, item);
        }
    }

    fn putDict(self: *Buf, comptime key_sig: u8, comptime val_sig: []const u8, items: anytype) void {
        self.padTo(4);
        self.putU32(@intCast(items.len));
        self.padTo(8);
        inline for (items) |item| {
            self.padTo(8);
            self.putValue(&.{key_sig}, item.key);
            self.putVariantValue(val_sig, item.value);
        }
    }

    fn putBody(self: *Buf, comptime sig: []const u8, values: anytype) void {
        const types = comptime sigTypes(sig);
        inline for (types, 0..) |t, i| {
            self.padTo(sigAlign(t));
            self.putValue(t, values[i]);
        }
    }

    pub fn serialise(self: *Buf, comptime msg_type: u8, comptime serial: u32, comptime headers: anytype, comptime body_sig: []const u8, body: anytype) void {
        self.putU8('l');
        self.putU8(msg_type);
        self.putU8(0);
        self.putU8(1);

        const body_len_off = self.off;
        self.off += 4;

        self.putU32(serial);

        const hdr_len_off = self.off;
        self.off += 4;

        const hdr_start = self.off;

        inline for (headers) |hdr| {
            self.padTo(8);
            self.putU8(hdr.code);
            self.putSignature(hdr.sig);
            self.padTo(sigAlign(hdr.sig));
            self.putValue(hdr.sig, hdr.val);
        }

        const hdr_len = self.off - hdr_start;
        std.mem.writeInt(u32, self.data[hdr_len_off..][0..4], @intCast(hdr_len), .little);

        self.padTo(8);
        const body_start = self.off;

        if (body_sig.len > 0) {
            putBody(self, body_sig, body);
        }

        const body_len = self.off - body_start;
        std.mem.writeInt(u32, self.data[body_len_off..][0..4], @intCast(body_len), .little);
    }

    fn putValue(self: *Buf, comptime sig: []const u8, val: anytype) void {
        const c = sig[0];
        if (c == 's' or c == 'o') {
            self.putString(val);
        } else if (c == 'u') {
            self.putU32(val);
        } else if (c == 'i') {
            self.putI32(val);
        } else if (c == 'g') {
            self.putSignature(val);
        } else if (c == 'v') {
            self.putVariantValue(val[0], val[1]);
        } else if (c == 'a') {
            if (sig[1] == '{') {
                self.putDict(sig[2], sig[3..sig.len - 1], val);
            } else {
                self.putArray(sig[1..], val);
            }
        } else if (c == '(') {
            const inner = comptime sigTypes(sig[1..sig.len - 1]);
            inline for (inner, 0..) |t, i| {
                self.padTo(sigAlign(t));
                self.putValue(t, val[i]);
            }
        }
    }

    fn sigAlign(comptime sig: []const u8) usize {
        const c = sig[0];
        if (c == 'y' or c == 'g') return 1;
        if (c == 'b' or c == 'i' or c == 'u' or c == 'h') return 4;
        if (c == 'n' or c == 'q') return 2;
        if (c == 'x' or c == 't' or c == 'd') return 8;
        if (c == 's' or c == 'o') return 4;
        if (c == 'v') return 1;
        if (c == 'a') {
            if (sig.len > 1 and sig[1] == '{') return 8;
            return @max(@as(usize, 4), sigAlign(sig[1..]));
        }
        if (c == '(') {
            const inner = comptime sigTypes(sig[1..sig.len - 1]);
            var max_align: usize = 1;
            inline for (inner) |t| {
                max_align = @max(max_align, sigAlign(t));
            }
            return max_align;
        }
        return 1;
    }

    fn sigTypes(comptime sig: []const u8) []const []const u8 {
        comptime {
            var result: [16][]const u8 = undefined;
            var count: usize = 0;
            var i: usize = 0;
            while (i < sig.len) {
                if (sig[i] == 'a') {
                    if (i + 1 < sig.len and sig[i + 1] == '{') {
                        const end = std.mem.indexOfScalarPos(u8, sig, i + 2, '}') orelse break;
                        result[count] = sig[i .. end + 1];
                        i = end + 1;
                    } else {
                        result[count] = sig[i .. i + 2];
                        i += 2;
                    }
                } else if (sig[i] == '(') {
                    const end = std.mem.indexOfScalarPos(u8, sig, i + 1, ')') orelse break;
                    result[count] = sig[i .. end + 1];
                    i = end + 1;
                } else {
                    result[count] = sig[i .. i + 1];
                    i += 1;
                }
                count += 1;
            }
            return result[0..count];
        }
    }
};

fn alignUp(off: usize, a: usize) usize {
    return (off + a - 1) & ~(a - 1);
}
